pantavisor.c

```diff
@@ -813,44 +813,47 @@ static void pv_remove(struct pantavisor *pv)
 
 static pv_state_t pv_shutdown(struct pantavisor *pv, shutdown_type_t t)
 {
-	char path[PATH_MAX];
-
 	if (!pv)
 		return PV_STATE_EXIT;
 
-	pv_log(INFO, "prepare %s...", shutdown_type_string(t));
-	wait_shell();
-
-	if (pv->state && pv_state_stop(pv->state))
-		pv_log(WARN, "stop error: ignoring due to %s",
+	pv_log(INFO, "preparing '%s'...", shutdown_type_string(t));
+	if (pv_config_get_system_init_mode() == IM_APPENGINE)
+		pv_log(INFO,
+		       "will not actually perform '%s' as we are in appengine mode",
 		       shutdown_type_string(t));
 
+	wait_shell();
+
 	if (REBOOT == t)
 		pv_wdt_start(pv);
 
-	// unmount storage
-	pv_paths_storage(path, PATH_MAX);
-	umount(path);
-	pv_fs_path_sync(path);
+	// stop childs leniently
+	pv_state_stop_lenient(pv->state);
+	ph_logger_stop_lenient();
 
-	sleep(5);
-	pv_log(INFO, "%s...", shutdown_type_string(t));
-	ph_logger_stop(pv);
-	pv_logserver_close();
-	ph_logger_close();
-	pv_ctrl_socket_close(pv->ctrl_fd);
+	// force stop childs
+	pv_state_stop_force(pv->state);
+	ph_logger_stop_force();
 
-	if (pv_config_get_system_init_mode() == IM_APPENGINE) {
-		pv_log(WARN,
-		       "closing application because of appengine init mode...");
-		pv_bootloader_remove();
-		pv_remove(pv);
-		goto out;
-	}
+	// at this point, we can shutdown if not in appengine
+	if (pv_config_get_system_init_mode() != IM_APPENGINE)
+		reboot(shutdown_type_reboot_cmd(t));
 
-	reboot(shutdown_type_reboot_cmd(t));
```

Comment:
is reboot never returning ... or do we wnat to run through the below functions in all cases?

Reply:
it's one of the things I was doubtful about. The idea is, in case of actual reboot or poweroff, we rely on those so we don't need to umount, close fds or freeup memory, which is only actually needed in case of appengine, as it is not triggering reboots or poweroffs for now